File: src/org/vaadin/hezamu/workouttracker/WorkoutEditorView.java
package org.vaadin.hezamu.workouttracker;

import java.util.Date;

import com.vaadin.data.Validator;
import com.vaadin.data.Validator.InvalidValueException;
import com.vaadin.shared.ui.label.ContentMode;
import com.vaadin.ui.Alignment;
import com.vaadin.ui.Button;
import com.vaadin.ui.ComboBox;
import com.vaadin.ui.DateField;
import com.vaadin.ui.GridLayout;
import com.vaadin.ui.Label;
import com.vaadin.ui.TextArea;
import com.vaadin.ui.TextField;
import com.vaadin.ui.themes.ValoTheme;

@SuppressWarnings("serial")
public class WorkoutEditorView extends GridLayout {
	final ComboBox activity;
	final DateField date;
	final TextField duration, avgHR, maxHR, calories;
	final Label title;
	final TextArea comment;
	final Button add, clear;

	public WorkoutEditorView() {
		super(2, 6);

		setSpacing(true);

		addComponent(title = new Label("New Workout"), 0, 0, 1, 0);
		title.setContentMode(ContentMode.HTML);
		title.addStyleName(ValoTheme.LABEL_H3);
		title.addStyleName(ValoTheme.LABEL_BOLD);
		title.setSizeUndefined();
		setComponentAlignment(title, Alignment.TOP_CENTER);

		addComponent(activity = new ComboBox("Activity"));
		activity.addItems(WorkoutPresenter.ACTIVITIES.keySet());
		activity.setRequired(true);
		activity.setInputPrompt("Required");

		addComponent(duration = new TextField("Duration"));
		duration.setRequired(true);
		duration.setInputPrompt("Required");
		duration.addValidator(new IntegerFieldValidator(1, 300));

		addComponent(date = new DateField("Date"));
		date.setValue(new Date());
		date.setRequired(true);
		date.setDateFormat("dd.MM.yyyy");
		date.addValidator(value -> {
			if (value instanceof Date && new Date().before((Date) value)) {
				throw new InvalidValueException("Date can't be in the future");
			}
		});

		addComponent(calories = new TextField("Calories"));
		calories.addValidator(new IntegerFieldValidator(0, 5000));

		addComponent(avgHR = new TextField("Average HR"));
		avgHR.addValidator(new IntegerFieldValidator(60, 200));

		addComponent(maxHR = new TextField("Max HR"));
		maxHR.addValidator(new IntegerFieldValidator(60, 200));

		addComponent(comment = new TextArea("Comment"), 0, 4, 1, 4);
		comment.setSizeFull();

		addComponent(add = new Button("Add"));
		add.addStyleName(ValoTheme.BUTTON_PRIMARY);
		add.setSizeFull();

		addComponent(clear = new Button("Clear"));
		clear.setSizeFull();
	}

	public void clearValues() {
		activity.setValue(null);
		duration.setValue("");
		date.setValue(new Date());
		calories.setValue("");
		avgHR.setValue("");
		maxHR.setValue("");
		comment.setValue("");
	}

	public int getDuration() {
		return getIntFieldValue(duration);
	}

	public int getCalories() {
		return getIntFieldValue(calories);
	}

	public double getAvgHR() {
		return getIntFieldValue(avgHR);
	}

	public double getMaxHR() {
		return getIntFieldValue(maxHR);
	}

	private int getIntFieldValue(TextField field) {
		try {
			return Integer.parseInt(field.getValue());
		} catch (NumberFormatException nfe) {
			return 0;
		}
	}

	private class IntegerFieldValidator implements Validator {
		private final int min, max;

		public IntegerFieldValidator(int min, int max) {
			this.min = min;
			this.max = max;
		}

		@Override
		public void validate(Object value) throws InvalidValueException {
			if (value == null || value.toString().isEmpty())
				return; // Empty values are caught with setRequired as needed

			int number;
			try {
				number = Integer.parseInt(value.toString());
			} catch (Exception e) {
				throw new InvalidValueException("Not an integer");
			}

			if (number < min || number > max) {
				throw new InvalidValueException("Must be between " + min
						+ " and " + max);
			}
		}
	}
}

File: src/org/vaadin/hezamu/workouttracker/WorkoutEditorView.scala
package org.vaadin.hezamu.workouttracker

import scala.util.Try
import vaadin.scala._
import vaadin.scala.Validator._
import com.vaadin.ui.themes.ValoTheme
import java.util.Date

class WorkoutEditorView extends GridLayout {
  columns = 2
  rows = 6
  spacing = true

  val title = add(new Label {
    value = "New Workout"
    sizeUndefined
    styleNames += ValoTheme.LABEL_H3 + " " + ValoTheme.LABEL_BOLD
  }, 0, 0, 1, 0, alignment = Alignment.TopCenter)
  title.contentMode = Label.ContentMode.Html

  val activity = add(new ComboBox {
    caption = "Activity"
    required = true
    inputPrompt = "Required"
    WorkoutPresenter.activities.foreach(addItem)
  })

  val duration = add(new ValidatedTextField("Duration", 1, 600, true))

  val date = add(new DateField {
    sizeFull
    caption = "Date"
    required = true
    value = new Date
    dateFormat = "dd.MM.yyyy"
  })

  date.validators += { v: Option[Any] =>
    v match {
      case Some(date: Date) if new Date().before(date) => Invalid(List("Date can't be in the future"))
      case _ => Valid
    }
  }

  val calories = add(new ValidatedTextField("Calories", 0, 5000))
  val avgHR = add(new ValidatedTextField("Average HR", 50, 250))
  val maxHR = add(new ValidatedTextField("Max HR", 50, 250))

  val comment = add(new TextArea {
    caption = "Comment"
    sizeFull
  }, 0, 4, 1, 4)

  val addButton = add(new Button {
    caption = "Add"
    sizeFull
    styleName = ValoTheme.BUTTON_PRIMARY
  })

  val clearButton = add(new Button {
    caption = "Clear"
    sizeFull
  })

  def clearFields {
    activity.value = None
    duration.value = ""
    date.value = new Date()
    calories.value = ""
    avgHR.value = ""
    maxHR.value = ""
    comment.value = ""
  }

  def getDuration = Try(Integer.parseInt(duration.value.getOrElse(""))) getOrElse 0
  def getCalories = Try(Integer.parseInt(calories.value.getOrElse(""))) getOrElse 0
  def getAvgHR = Try(Integer.parseInt(avgHR.value.getOrElse(""))) getOrElse 0
  def getMaxHR = Try(Integer.parseInt(maxHR.value.getOrElse(""))) getOrElse 0
}

// Vaadin supports JSR 303 Java Bean Validation API as well
class ValidatedTextField(_caption: String, min: Int, max: Int, _required: Boolean = false) extends TextField {
  this.caption = _caption
  this.required = _required

  validators += { v =>
    v match {
      case Some(s: String) if s.nonEmpty =>
        Try(Integer.parseInt(s)) map {
          case num if num >= min && num <= max => Valid
          case _ => Invalid(List(s"Must be between $min and $max"))
        } getOrElse Invalid(List("Not an integer"))
      case _ => Valid // Empty values are caught with setRequired as needed
    }
  }
}
